awriter/signer.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2019 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package awriter
+
+import (
+	"archive/tar"
+	"io"
+
+	"github.com/mendersoftware/mender-artifact/artifact"
+	"github.com/pkg/errors"
+)
+
+var ErrAlreadyExistingSignature = errors.New("Will not overwrite existing signature")
+
+// Special fast-track to just sign, nothing else. This skips all the expensive
+// and complicated repacking, and simply adds the manifest.sig file.
+func SignExisting(src io.Reader, dst io.Writer, key []byte, overwrite bool) error {
+	var foundManifest bool
+	rTar := tar.NewReader(src)
+	wTar := tar.NewWriter(dst)
+	buf := make([]byte, 1048576)
+	for {
+		header, err := rTar.Next()
+		if err == io.EOF {
+			break
+		} else if (err != nil) {
+			return errors.Wrap(err, "Could not read tar header")
+		}
+
+		switch header.Name {
+		case "manifest":
+			err = signManifestAndOutputSignature(header, rTar, wTar, key)
+			if err != nil {
+				return err
+			}
+			foundManifest = true
+			continue
+		case "manifest.sig":
+			if overwrite {
+				continue
+			} else {
+				return ErrAlreadyExistingSignature
+			}
+		}
+
+		err = wTar.WriteHeader(header)
+		if err != nil {
+			return errors.Wrap(err, "Could not write tar header")
+		}
+
+		read, err := rTar.Read(buf)
+		if err != nil && err != io.EOF {
+			return errors.Wrap(err, "Could not read tar file entry")
+		} else if read == 0 {
+			continue
+		}
+
+		written, err := wTar.Write(buf[:read])
+		if err != nil {
+			return errors.Wrap(err, "Could not write tar file entry")
+		} else if written != read {
+			return errors.New("Not able to write entire tar file entry")
+		}
+	}
+
+	err := wTar.Close()
+	if err != nil {
+		return errors.Wrap(err, "Could not finalize tar archive")
+	}
+
+	if !foundManifest {
+		return errors.New("`manifest` not found. Corrupt Artifact?")
+	}
+
+	return nil
+}
+
+func signManifestAndOutputSignature(header *tar.Header, src *tar.Reader, dst *tar.Writer, key []byte) error {
+	signer := artifact.NewSigner(key)
+	buf := make([]byte, header.Size)
+	read, err := src.Read(buf)
+	if err != nil && err != io.EOF {
+		return errors.Wrap(err, "Could not read manifest")
+	} else if int64(read) != header.Size {
```

Comment:
It's actually `Read` which is inconsistent in this respect: It doesn't allow reading more than 2GiB at a time. Not that most people would need that, but it does create this need for casting, since tar clearly needs to allow files > 2GiB. This is a fairly common casting need in Go. You have the same problem with `stat` and its size.

Reply:
Nice, did not know!